JENKINS-60722 add maxDaysOld parameter to restrict age of multibranches

dsteiert opened the pull request:
If there is no filter set and the configuration is changed or a new Jenkins is started (after an If there is no filter set and the configuration is changed or a new Jenkins is started (after an upgrade), all multi-branch pipeline jobs are executed at the same time. To reduce the amount of jobs running at the same time (especially with multiple projects, repositories and branches), a parameter is introduced, which restricts relevant branches, tags and pull requests according to the age of the head commit.

daniel-beck commented:
Thanks for your pull request. Please note that this plugin is not very actively maintained as I consider it "done", so responses may be delayed.

> especially in the future with Team Pipelines

Could you clarify what this is in reference to?

dsteiert commented:
"Team Pipelines" refers to our internal company efforts to provide pipelines for multiple teams that only differ in configuration changes. Every team will get access to a Jenkins that is completely set up and configured out of the box already (Pipeline as Code, Infrastructure as Code aso.), where they only have to log in to check logs. This is nothing special and I assume basically everybody in every company has a setup where teams use multibranch plugins or at least build multiple projects with a few branches regularly. 
Due to a lot of features, bugfixes and maintenance, there will be a lot of branches per repo, while there also are a lot of repos per project/team. If there now is some form of outage where for any reason at all (e.g. stateless Jenkins, data-loss, data-corruption, seed-job rerun, plugins/jobs reconfigured), all attached branches/tags/PRs are rebuilt. Even if those branches are older than cutting edge development. 

This can also happen in general integration pipelines, but there is even more demand for those (above mentioned) team-pipelines that create high load and need fast feedback to possible during the entire day.

daniel-beck commented:
Thanks, that clarifies that no readers of the initial PR comment were supposed to fully understand it.

----

Notable references:

GitHub BS sometimes returns 0 to indicate invalid states, so perhaps we should treat that value as "always outdated"?
* https://github.com/jenkinsci/github-branch-source-plugin/blob/28603b8545f0de2876e04eb3bea424317ebe2181/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMProbe.java#L110-L136

Will not work with Mercurial at all:
* https://github.com/jenkinsci/mercurial-plugin/blob/8672947da2ada10a480dba9b1c4e5348a0a0c304/src/main/java/hudson/plugins/mercurial/MercurialSCMSource.java#L455

Gitea also uses 0 as an error value:
* https://github.com/jenkinsci/gitea-plugin/blob/796fa5953c9e56cc26a7fe07a9fcba4f88f2cd21/src/main/java/org/jenkinsci/plugin/gitea/GiteaSCMSource.java#L530
… and as the value for non-error situations too:
* https://github.com/jenkinsci/gitea-plugin/blob/796fa5953c9e56cc26a7fe07a9fcba4f88f2cd21/src/main/java/org/jenkinsci/plugin/gitea/GiteaSCMFileSystem.java#L84

GitLab looks reasonable but also uses 0 as a special error value:
* https://github.com/jenkinsci/gitlab-branch-source-plugin/blob/4034bb6b230c2ee19ab11dbb64e53f210831c0ab/src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMFileSystem.java#L52
* https://github.com/jenkinsci/gitlab-branch-source-plugin/blob/4034bb6b230c2ee19ab11dbb64e53f210831c0ab/src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java#L694

----

In light of the above, I suggest this be rewritten to treat `0` values returned from the API as an error / undefined value which will never be used while the feature is active. (Of course, given that 0 is a 1970 timestamp, that's pretty much automatic anyway so 🤷‍♂ )

Additionally, this needs an inline HTML help file explaining how it (generally) works, including that some SCM implementations may not support this option, and what exactly the "day boundaries" are. Are we rounding down, up, to nearest full day, and does the current time matter?

dsteiert commented:
Sorry for that. I updated the commit message accordingly and will adapt the PR message. It makes sense to define the `probe.lastModified()` return value `0` as an error/default state. I added the changes accordingly. The PR now also includes the added inline HTML file with helpful instructions.

Thanks!

dsteiert commented:
The `InlineDefinitionMultiBranchProjectFactory` helpfile for `maxDaysOld` was missing. I added it there as well.

dsteiert commented:
I just saw something: If no {{Markerfile}} is specified, no branches and nothing is built. I understand that the option is called {{Common pipeline definition for markerfile}}, but not actually filling in the field for the {{Markerfile}} (which is not restricted, anybody can do it without verification), basically makes the whole job worthless. Nothing will ever be built. 
Is this wanted behaviour, should it be verified that it is not empty or should it behave as if a {{Markerfile}} was present?
I could create a ticket and prepare a solution in case it is not wanted behaviour.

dsteiert commented:
I had looked at the basic-branch-build-strategies plugin before but hadn't seen what I needed/wanted. After a lot more research and digging we got one of the strategies, `skipInitialBuildOnFirstBranchIndexing` working also with the Job-DSL. Together with the functionality of the `github-scm-filter-aged-refs` plugin and the `tagBuildStrategy` of the `basic-branch-build-strategies` plugin, we can probably solve our issues.
This basically enables us to not build every single branch when the Jenkins is recreated or when there is data corruption and we can filter branches and tags out that are too old.

Thanks for all your help and the time you took to review and look for solutions! I think we could close this pull request with the other solutions in place.
